clients/mapiclient/msqldump.1
.\" SPDX-License-Identifier: MPL-2.0
.\"
.\" This Source Code Form is subject to the terms of the Mozilla Public
.\" License, v. 2.0.  If a copy of the MPL was not distributed with this
.\" file, You can obtain one at https://mozilla.org/MPL/2.0/.
.\"
.\" For copyright information, see the file debian/copyright.
.\"
.TH MSQLDUMP 1 MonetDB
.SH NAME
msqldump \- dump a MonetDB/SQL database
.SH SYNOPSIS
.B msqldump
[
.I options
] [
.I dbname
]
.SH DESCRIPTION
MonetDB is a database management system that is developed from a
main-memory perspective with use of a fully decomposed storage model,
automatic index management, extensibility of data types and search
accelerators, and an SQL front end.
.PP
.I Msqldump
is the program to dump an MonetDB/SQL database.
The dump can be used to populate a new MonetDB/SQL database.
.PP
Before
.I msqldump
starts parsing command line options, it reads a configuration file.
If the environment variable
.B DOTMONETDBFILE
is set and not empty, it reads the file pointed to by that variable.
If set but empty, no configuration file is read.
If unset,
.I msqldump
first looks for a file
.I .monetdb
in the current working directory, and if that doesn't exist, it
looks for a file
.I monetdb
in the XDG configuration directory
.RB ( $XDG_CONFIG_HOME
which defaults to
.B $HOME/.config
if not set), and finally for a file
.I .monetdb
in the current user's home directory.
This file can contain defaults for the flags
.BR user ,
.BR password ,
.BR host ,
and
.BR port .
To disable reading the
.I .monetdb
file, set the variable
.B DOTMONETDBFILE
to the empty string in the environment.
.SH OPTIONS
.TP
\fB\-\-help\fP (\fB\-?\fP)
Print usage information and exit.
.TP
\fB\-\-database=\fP\fIdatabase\fP (\fB\-d\fP \fIdatabase\fP)
Specify the name of the database to connect to.
The
.B \-d
can be omitted if it is the last option.
.TP
\fB\-\-host=\fP\fIhostname\fP (\fB\-h\fP \fIhostname\fP)
Specify the name of the host on which the server runs (default:
.BR localhost ).
.TP
\fB\-\-port=\fP\fIportnr\fP (\fB\-p\fP \fIportnr\fP)
Specify the portnumber of the server (default: 50000).
.TP
\fB\-\-user\fP\fB=\fP\fIuser\fP (\fB\-u\fP \fIuser\fP)
Specify the user to connect as.
If this flag is absent, the client will ask for a user name.
.TP
\fB\-\-describe\fP (\fB\-D\fP)
Only dump the database schema.
.TP
\fB\-\-inserts\fP (\fB\-N\fP)
When dumping the table data, use INSERT INTO statements, rather than
COPY INTO + CSV values.
INSERT INTO statements are more portable, and necessary when the
load of the dump is processed by e.g. a JDBC application.
When the dump is to be used to feed back into MonetDB, COPY
INTO is preferred since that is processed more efficiently.
.TP
\fB\-\-noescape\fP (\fB\-e\fP)
When dumping the table data, use the NO ESCAPE option on the COPY INTO
query.
.TP
\fB\-\-functions\fP (\fB\-f\fP)
Only dump function definitions.
.TP
\fB\-\-table=[\fIschema\fP\fB.\fP]\fItable\fP (\fB\-t\fP [\fIschema\fP\fB.\fP]\fItable\fP)
Only dump the specified table.
If
.I schema
is not specified, the user's current schema is used.
When either
.I schema
or
.I table
contains percent characters, all tables matching the (SQL) search
pattern are dumped.
.TP
\fB\-\-output=\fP\fIfilename\fP (\fB\-o\fP \fIfilename\fP)
Write the dump to the specified file.
If
.I filename
contains a recognized compression scheme suffix, the file will be
compressed using that scheme.
.TP
\fB\-\-outputdir=\fP\fIdirectory\fP (\fB\-O\fP \fIdirectory\fP)
Write the dump to the file
.B dump.sql
in the specified directory which is created if it does not exist.
The data of the tables will be stored in separate CSV files in the
directory.
If the
.B \-\-compression
option is used, the CSV files will be compressed using the specified
compression scheme.
If both the 
.B \-\-outputdir
option and the
.B \-\-output
option are specified, the last one is used.
The
.B \-\-outputdir
option is not compatible with the
.B \-\-inserts
option.
.TP
\fB\-\-compression=\fP\fIextension\fP (\fB\-x\fP \fIextension\fP)
Compress the CSV files produced with the
.B \-\-outputdir
option using the specified compression scheme.
Depending on how the program is compiled, the following compression
schemes are recognized:
.BR gz ,
.BR bz2 ,
.BR xz ,
and
.BR lz4 .
.TP
\fB\-\-quiet\fP (\fB\-q\fP)
Don't print the welcome message.
.TP
\fB\-\-Xdebug\fP (\fB\-X\fP)
Trace network interaction between
.I mclient
and the server.
.SH SEE ALSO
.IR mclient (1),
.IR mserver5 (1)


ChangeLog
# ChangeLog file for devel
# This file is updated with Maddlog

* Thu Feb 19 2026 Sjoerd Mullender <sjoerd@acm.org>
- Changed the BATloop macro so that it now has a BATiter pointer as
  first argument.



clients/mapiclient/msqldump.c
/*
 * SPDX-License-Identifier: MPL-2.0
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0.  If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * For copyright information, see the file debian/copyright.
 */

#include "monetdb_config.h"
#ifdef HAVE_GETOPT_H
#include "getopt.h"
#endif
#include "mapi.h"
#include <unistd.h>
#include <sys/stat.h>
#include <string.h>
#include <time.h>

#include "stream.h"
#include "msqldump.h"
#include "mprompt.h"
#include "mutils.h"		/* mercurial_revision */
#include "dotmonetdb.h"

static _Noreturn void usage(const char *prog, int xit);

static void
usage(const char *prog, int xit)
{
	fprintf(stderr, "Usage: %s [ options ] [ dbname ]\n", prog);
	fprintf(stderr, "\nOptions are:\n");
	fprintf(stderr, " -h hostname | --host=hostname    host to connect to\n");
	fprintf(stderr, " -p portnr   | --port=portnr      port to connect to\n");
	fprintf(stderr, " -u user     | --user=user        user id\n");
	fprintf(stderr, " -d database | --database=database  database to connect to\n");
	fprintf(stderr, " -o filename | --output=filename  write dump to filename\n");
	fprintf(stderr, " -O dir      | --outputdir=dir    write multi-file dump to dir\n");
	fprintf(stderr, " -x ext      | --compression=ext  compression method ext for multi-file dump\n");
	fprintf(stderr, " -f          | --functions        dump functions\n");
	fprintf(stderr, " -t table    | --table=table      dump a database table\n");
	fprintf(stderr, " -D          | --describe         describe database\n");
	fprintf(stderr, " -N          | --inserts          use INSERT INTO statements\n");
	fprintf(stderr, " -e          | --noescape         use NO ESCAPE\n");
	fprintf(stderr, " -q          | --quiet            don't print welcome message\n");
	fprintf(stderr, " -X          | --Xdebug           trace mapi network interaction\n");
	fprintf(stderr, " -?          | --help             show this usage message\n");
	fprintf(stderr, "--functions and --table are mutually exclusive\n");
	fprintf(stderr, "--output and --outputdir are mutually exclusive\n");
	fprintf(stderr, "--inserts and --outputdir are mutually exclusive\n");
	fprintf(stderr, "--compression only has effect with --outputdir\n");
	exit(xit);
}

int
#ifdef _MSC_VER
wmain(int argc, wchar_t **wargv)
#else
main(int argc, char **argv)
#endif
{
	int port = 0;
	const char *user = NULL;
	const char *passwd = NULL;
	const char *host = NULL;
	const char *dbname = NULL;
	const char *output = NULL;
	const char *outputdir = NULL;
	const char *ext = NULL;
	DotMonetdb dotfile = {0};
	bool trace = false;
	bool describe = false;
	bool functions = false;
	bool useinserts = false;
	bool noescape = false;
	int c;
	Mapi mid;
	bool quiet = false;
	stream *out;
	bool user_set_as_flag = false;
	const char *table = NULL;
	static struct option long_options[] = {
		{"host", 1, 0, 'h'},
		{"port", 1, 0, 'p'},
		{"database", 1, 0, 'd'},
		{"output", 1, 0, 'o'},
		{"outputdir", 1, 0, 'O'},
		{"compression", 1, 0, 'x'},
		{"describe", 0, 0, 'D'},
		{"functions", 0, 0, 'f'},
		{"table", 1, 0, 't'},
		{"inserts", 0, 0, 'N'},
		{"noescape", 0, 0, 'e'},
		{"Xdebug", 0, 0, 'X'},
		{"user", 1, 0, 'u'},
		{"quiet", 0, 0, 'q'},
		{"version", 0, 0, 'v'},
		{"help", 0, 0, '?'},
		{0, 0, 0, 0}
	};
#ifdef _MSC_VER
	char **argv = malloc((argc + 1) * sizeof(char *));
	if (argv == NULL) {
		fprintf(stderr, "cannot allocate memory for argument conversion\n");
		exit(1);
	}
	for (int i = 0; i < argc; i++) {
		if ((argv[i] = utf16toutf8(wargv[i])) == NULL) {
			fprintf(stderr, "cannot convert argument to UTF-8\n");
			exit(1);
		}
	}
	argv[argc] = NULL;
#endif

	parse_dotmonetdb(&dotfile);
	user = dotfile.user;
	passwd = dotfile.passwd;
	dbname = dotfile.dbname;
	host = dotfile.host;
	port = dotfile.port;

	while ((c = getopt_long(argc, argv, "h:p:d:o:O:x:Dft:NeXu:qv?", long_options, NULL)) != -1) {
		switch (c) {
		case 'u':
			user = optarg;
			user_set_as_flag = true;
			break;
		case 'h':
			host = optarg;
			break;
		case 'p':
			assert(optarg != NULL);
			port = atoi(optarg);
			break;
		case 'd':
			dbname = optarg;
			break;
		case 'o':
			output = optarg;
			outputdir = NULL;
			break;
		case 'O':
			outputdir = optarg;
			output = NULL;
			break;
		case 'x':
			ext = optarg;
			break;
		case 'D':
			describe = true;
			break;
		case 'N':
			useinserts = true;
			break;
		case 'e':
			noescape = true;
			break;
		case 'f':
			if (table)
				usage(argv[0], -1);
			functions = true;
			break;
		case 't':
			if (table || functions)
				usage(argv[0], -1);
			table = optarg;
			break;
		case 'q':
			quiet = true;
			break;
		case 'X':
			trace = true;
			break;
		case 'v': {
			printf("msqldump, the MonetDB interactive database "
			       "dump tool, version %s", MONETDB_VERSION);
#ifdef MONETDB_RELEASE
			printf(" (%s)", MONETDB_RELEASE);
#else
			const char *rev = mercurial_revision();
			if (strcmp(rev, "Unknown") != 0)
				printf(" (hg id: %s)", rev);
#endif
			printf("\n");
			destroy_dotmonetdb(&dotfile);
			return 0;
		}
		case '?':
			/* a bit of a hack: look at the option that the
			   current `c' is based on and see if we recognize
			   it: if -? or --help, exit with 0, else with -1 */
			usage(argv[0], strcmp(argv[optind - 1], "-?") == 0 || strcmp(argv[optind - 1], "--help") == 0 ? 0 : -1);
		default:
			usage(argv[0], -1);
		}
	}

	if ((output != NULL || useinserts) && outputdir) {
		usage(argv[0], -1);
	}

	if (optind == argc - 1)
		dbname = argv[optind];
	else if (optind != argc)
		usage(argv[0], -1);

	/* when config file would provide defaults */
	if (user_set_as_flag)
		passwd = NULL;

	if(dbname == NULL){
		printf("msqldump, please specify a database\n");
		usage(argv[0], -1);
	}
	char *user_allocated = NULL;
	if (user == NULL) {
		user_allocated = simple_prompt("user", BUFSIZ, 1, prompt_getlogin());
		user = user_allocated;
	}
	char *passwd_allocated = NULL;
	if (passwd == NULL) {
		passwd_allocated = simple_prompt("password", BUFSIZ, 0, NULL);
		passwd = passwd_allocated;
	}

	if (dbname != NULL && strchr(dbname, ':') != NULL) {
		mid = mapi_mapiuri(dbname, user, passwd, "sql");
	} else {
		mid = mapi_mapi(host, port, user, passwd, "sql", dbname);
	}
	free(user_allocated);
	user_allocated = NULL;
	free(passwd_allocated);
	passwd_allocated = NULL;
	user = NULL;
	passwd = NULL;
	dbname = NULL;
	if (mid == NULL) {
		fprintf(stderr, "failed to allocate Mapi structure\n");
		exit(2);
	}
	if (mapi_error(mid)) {
		mapi_explain(mid, stderr);
		exit(2);
	}
	mapi_set_time_zone(mid, 0);
	mapi_reconnect(mid);
	if (mapi_error(mid)) {
		mapi_explain(mid, stderr);
		exit(2);
	}
	if (!quiet) {
		const char *motd = mapi_get_motd(mid);

		if (motd)
			fprintf(stderr, "%s", motd);
	}
	mapi_trace(mid, trace);
	mapi_cache_limit(mid, -1);

	if (output) {
		out = open_wastream(output);
	} else if (outputdir) {
		size_t fnl = strlen(outputdir) + 10;
		if (ext)
			fnl += strlen(ext) + 1;
		char *fn = malloc(fnl);
		if (fn == NULL) {
			fprintf(stderr, "malloc failure\n");
			exit(2);
		}
		if (MT_mkdir(outputdir) == -1 && errno != EEXIST) {
			perror("cannot create output directory");
			exit(2);
		}
		snprintf(fn, fnl, "%s%cdump.sql", outputdir, DIR_SEP);
		out = open_wastream(fn);
		free(fn);
		(void) ext;
	} else {
		out = stdout_wastream();
	}
	if (out == NULL) {
		if (output)
			fprintf(stderr, "cannot open file: %s: %s\n",
					output, mnstr_peek_error(NULL));
		else if (outputdir)
			fprintf(stderr, "cannot open file: %s%cdump.sql: %s\n",
					outputdir, DIR_SEP, mnstr_peek_error(NULL));
		else
			fprintf(stderr, "failed to allocate stream: %s\n",
					mnstr_peek_error(NULL));
		exit(2);
	}
	if (!quiet) {
		char buf[27];
		time_t t = time(0);
		char *p;

#ifdef HAVE_CTIME_R3
		ctime_r(&t, buf, sizeof(buf));
#else
#ifdef HAVE_CTIME_R
		ctime_r(&t, buf);
#else
		strcpy_len(buf, ctime(&t), sizeof(buf));
#endif
#endif
		if ((p = strrchr(buf, '\n')) != NULL)
			*p = 0;

		mnstr_printf(out,
			     "-- msqldump version %s", MONETDB_VERSION);
#ifdef MONETDB_RELEASE
		mnstr_printf(out, " (%s)", MONETDB_RELEASE);
#else
		const char *rev = mercurial_revision();
		if (strcmp(rev, "Unknown") != 0)
			mnstr_printf(out, " (hg id: %s)", rev);
#endif
		mnstr_printf(out, " %s %s%s\n",
			     describe ? "describe" : "dump",
			     functions ? "functions" : table ? "table " : "database",
			     table ? table : "");
		dump_version(mid, out, "-- server:");
		mnstr_printf(out, "-- %s\n", buf);
	}
	if (functions) {
		mnstr_printf(out, "START TRANSACTION;\n");
		c = dump_functions(mid, out, true, NULL, NULL, NULL);
		mnstr_printf(out, "COMMIT;\n");
	} else if (table) {
		mnstr_printf(out, "START TRANSACTION;\n");
		c = dump_table(mid, NULL, table, out, outputdir, ext, describe, true, useinserts, false, noescape, true);
		mnstr_printf(out, "COMMIT;\n");
	} else
		c = dump_database(mid, out, outputdir, ext, describe, useinserts, noescape);
	mnstr_flush(out, MNSTR_FLUSH_DATA);

	mapi_destroy(mid);
	if (mnstr_errnr(out) != MNSTR_NO__ERROR) {
		fprintf(stderr, "%s: %s\n", argv[0], mnstr_peek_error(out));
		c = 1;
	}

	close_stream(out);

	destroy_dotmonetdb(&dotfile);

	return c;
}
